### PR TITLE
[docs] fix broken internal references

### PR DESCRIPTION
**Pull request by maxpumperla**
Signed-off-by: Max Pumperla <max.pumperla@googlemail.com>

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by simon-mo — APPROVED** [review]

**Review by DmitriGekhtman — APPROVED** [review]

File: doc/source/data/creating-datasets.rst [hunk 1]
```diff
@@ -533,11 +533,11 @@ Load image data stored as individual files using :py:class:`~ray.data.datasource
 
 .. _datasets_custom_datasource:
 
-------------------
-Custom Datasources
-------------------
+-------------------
+Custom Data Sources
```
Comment (clarkzinzow) [inline doc/source/data/creating-datasets.rst:537]: We've been using "datasource" as a single-word concept paired with the `Datasource` abstraction, and although I'm actually in favor of changing this to a "data source" concept + a `DataSource` abstraction, I'd consider that to be out-of-scope for this reference fix PR.
```suggestion
Custom Datasources
```

File: doc/source/data/creating-datasets.rst [hunk 2]
```diff
@@ -533,11 +533,11 @@ Load image data stored as individual files using :py:class:`~ray.data.datasource
...
+-------------------
 
-Datasets can read and write in parallel to `custom datasources <package-ref.html#custom-datasource-api>`__ defined in Python.
+Datasets can read and write in parallel to :ref:`custom data sources <data_source_api>` defined in Python.
```
Comment (clarkzinzow) [inline doc/source/data/creating-datasets.rst:540]: Ditto
```suggestion
Datasets can read and write in parallel to :ref:`custom datasources <data_source_api>` defined in Python.
```

File: doc/source/data/creating-datasets.rst [hunk 3]
```diff
@@ -533,11 +533,11 @@ Load image data stored as individual files using :py:class:`~ray.data.datasource
 
 .. _datasets_custom_datasource:
 
-------------------
+-------------------
 Custom Datasources
-------------------
+-------------------
```
Comment (richardliaw) [inline doc/source/data/creating-datasets.rst:538]:
```suggestion
------------------
```

File: doc/source/data/creating-datasets.rst [hunk 4]
```diff
@@ -533,11 +533,11 @@ Load image data stored as individual files using :py:class:`~ray.data.datasource
 
 .. _datasets_custom_datasource:
 
-------------------
+-------------------
```
Comment (richardliaw) [inline doc/source/data/creating-datasets.rst:536]:
```suggestion
------------------
```

**Review by pcmoritz — APPROVED** [review]
Thanks! The references are working now 🎉

**Pull request merged**